dm/syncer/binlogstream/streamer_controller.go

```diff
@@ -278,37 +286,221 @@ func (c *StreamerController) resetReplicationSyncer(tctx *tcontext.Context, loca
+
+				return nil, 0, pb.ErrorOp_InvalidErrorOp, err
+			}
+		}
+
+		// fake rotate. binlog recorder should handle it
+		if c.lastEventFromUpstream.Header.LogPos == 0 {
+			event = c.lastEventFromUpstream
+			c.lastEventFromUpstream = nil
+			return
+		}
+
+		startPos := mysql.Position{
+			Name: c.currBinlogFile,
+			Pos:  c.lastEventFromUpstream.Header.LogPos - c.lastEventFromUpstream.Header.EventSize,
+		}
+		cmp := binlog.ComparePosition(startPos, frontOp.pos)
+		switch cmp {
+		// when upstream event is earlier than any injected op.
+		case -1:
+			event = c.lastEventFromUpstream
+			c.lastEventFromUpstream = nil
+			break LOOP
+		// when upstream event is at the same location as injected op.
+		case 0:
+			// TODO: check it's DDL
+			switch frontOp.op {
+			case pb.ErrorOp_Skip:
+				// skipped event and op should be sent to caller, to let schema
+				// tracker and checkpoint work
+				event = c.lastEventFromUpstream
+				c.lastEventFromUpstream = nil
+				c.streamModifier.next()
+				op = pb.ErrorOp_Skip
+				break LOOP
+			case pb.ErrorOp_Replace:
+				// this op has been totally consumed, move to next op and also delete
+				// lastEventFromUpstream because it's Replace.
+				if c.streamModifier.nextEventInOp == len(frontOp.events) {
+					c.lastEventFromUpstream = nil
+					c.streamModifier.next()
+					continue
+				}
+
+				event = frontOp.events[c.streamModifier.nextEventInOp]
+				c.streamModifier.nextEventInOp++
+
+				// for last event in Replace, we change the LogPos and EventSize
+				// to imitate the real event. otherwise, we use the LogPos from
+				// startPosition, zero EventSize and increase suffix.
+				if c.streamModifier.nextEventInOp == len(frontOp.events) {
+					event.Header.LogPos = c.lastEventFromUpstream.Header.LogPos
+					event.Header.EventSize = c.lastEventFromUpstream.Header.EventSize
+					// TODO: for last event we should forward GTID set? let caller
+					// do it unless we merge location recorder
+					suffix = 0
+				} else {
+					event.Header.LogPos = startPos.Pos
+					suffix = c.streamModifier.nextEventInOp
+				}
+				op = pb.ErrorOp_Replace
+				// TODO: currently we let caller to modify Suffix, after location recorder
+				// binlog streamer itself can maintain it
+				break LOOP
+			case pb.ErrorOp_Inject:
+				// this op has been totally consumed, move to next op and return
+				// original upstream event
+				if c.streamModifier.nextEventInOp == len(frontOp.events) {
+					c.streamModifier.next()
+					event = c.lastEventFromUpstream
+					c.lastEventFromUpstream = nil
+					break LOOP
+				}
+
+				event = frontOp.events[c.streamModifier.nextEventInOp]
+				c.streamModifier.nextEventInOp++
+
+				event.Header.LogPos = startPos.Pos
+				suffix = c.streamModifier.nextEventInOp
+				op = pb.ErrorOp_Inject
+				break LOOP
+			default:
+				c.logger.DPanic("invalid error handle op", zap.Stringer("op", frontOp.op))
+			}
+		// when upstream event is later than any injected op. Apart from Inject,
```

Comment:
Should it be "...later than any events in the frontOp"?

Reply:
some op don't not have events, like SkipOp

Reply:
A bit obscure here😵‍💫 `case 1` means the position of the `frontOp` is smaller than the binlog event, doesn't it? But the "**any** injected op" baffles me. Does the "any injected op" means "all injected op"?

Reply:
Oh, I should say frontOp (all op that are not consumed)

Reply:
https://github.com/pingcap/tiflow/pull/5973/commits/d48c3afcc423e6e9492c91016c2513630ea69d5a